File: FMCC/Scripts/app/controllers/AdminController.js
angular.module("fmccwebportal").controller("adminController", function ($scope, $http) {

    var role = angular.element(".role-form-elem");

    var module = angular.element(".module-form-elem");

    role.select2({
        allowClear: true,
        placeholder: "Select a user role"
    });

    module.select2({
        allowClear: true,
        placeholder: "Select a module"
    });

    $http.get("/api/admin/getrolelist").error(function (res) {
        console.log(res);
    }).success(function (res) {
        if (res.okay) {
            role.select2({
                data: res.model,
                placeholder: "Select a user role"
            });
        } else {
            console.log(res.message);
        }
    });

    $http.get("/api/admin/getmodulelist").error(function (res) {
        console.log(res);
    }).success(function (res) {
        if (res.okay) {
            module.select2({
                data: res.model,
                placeholder: "Select a module"
            });
        } else {
            console.log(res.message);
        }
    });

    $scope.loadMenuItem = function () {
        if (role.val().trim() == "") {
            alert("Please select a role!");
        } else if (module.val().trim() == "") {
            alert("Please select a module!");
        } else {
            var url = "/api/admin/getmenulist?role=" + role.val() + "&moduleId=" + module.val();
            $http.get(url).error(function (res) {
                console.log(res);
            }).success(function (res) {
                if (res.okay) {
                    $scope.menuList = res.model;
                } else {
                    console.log(res.message);
                }
            });
        }
    }

    $scope.SaveMenuItem = function () {
        if ($scope.menuList.length) {
            $scope.menuList.RoleId = role.val();
            $http.post("/api/admin/setmenulist", $scope.menuList).error(function (res) {
                console.log(res);
                role.val("").trigger("change");
                module.val("").trigger("change");
            }).success(function (res) {
                role.val("").trigger("change");
                module.val("").trigger("change");
                if (res.okay) {
                    $scope.menuList = [];
                   
                } else {
                    alert(res.message)
                }
            });
        }
    }

    $scope.SetChecked = function (item, menuList) {
        if (item.Parent == 0) {
            if (item.IsAuthorized) {
                menuList.forEach(function (menu) {
                    menu.IsAuthorized = false;
                });
            } else {
                menuList.forEach(function (menu) {
                    menu.IsAuthorized = true;
                });
            }
        } else {
            if (item.IsAuthorized) {
                item.IsAuthorized = false;
                var hasSelectedChild = false;
                menuList.forEach(function (menu) {
                    if (menu.Parent != 0) {
                        hasSelectedChild = menu.IsAuthorized ? true : hasSelectedChild;
                    }
                });
                if (hasSelectedChild) {
                    menuList.forEach(function (menu) {
                        if (menu.Parent == 0) {
                            menu.IsAuthorized = true;
                        }
                    });
                } else {
                    menuList.forEach(function (menu) {
                        if (menu.Parent == 0) {
                            menu.IsAuthorized = false;
                        }
                    });
                }
            } else {
                item.IsAuthorized = true;
                menuList.forEach(function (menu) {
                    if (menu.Parent == 0) {
                        menu.IsAuthorized = true;
                    }
                });
            }
        }
    };

});

File: FMCC/Controllers/ServiceControllers/AdminController.cs
using System;
using System.Linq;
using Fmcc.Models;
using System.Web.Http;
using Fmcc.Models.EntityDataModel;
using System.Collections.Generic;


namespace Fmcc.Controllers.ServiceControllers
{

    [RoutePrefix("api/admin")]
    public class AdminController : ApiController
    {

        readonly log4net.ILog logger = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
        private FMCCDataContext context;

        public AdminController()
        {
            context = new FMCCDataContext();
        }

        [HttpGet]
        [Route("getrolelist")]
        public Output GetRoleList()
        {
            Output output = new Output();
            try
            {
                var list = context.Roles.Select(e => new { id = e.Id, text = e.Name }).ToList();
                output.okay = true;
                output.model = list;
                output.message = string.Empty;
            }
            catch (Exception exception)
            {
                output.okay = false;
                output.model = null;
                output.message = exception.Message;
                logger.Error(exception.Message);
            }
            return output;
        }

        [HttpGet]
        [Route("getmodulelist")]
        public Output GetModuleList()
        {
            Output output = new Output();
            try
            {
                var list = context.Modules.Select(e => new { id = e.Id, text = e.Name }).ToList();
                output.okay = true;
                output.model = list;
                output.message = string.Empty;
            }
            catch (Exception exception)
            {
                output.okay = false;
                output.model = null;
                output.message = exception.Message;
                logger.Error(exception.Message);
            }
            return output;
        }

        [HttpGet]
        [Route("getmenulist")]
        public object GetMenuList(int role = 0, int moduleId = 0)
        {
            Output output = new Output();
            try
            {
                if (context.MenuAuthentications.Where(w => w.RoleId == role && w.ModuleId == moduleId).Any())
                {
                    output.model = context.MenuAuthentications
                        .Where(w => w.RoleId == role && w.ModuleId == moduleId)
                        .Join(context.Menus, menuauth => menuauth.MenuId, menu => menu.Id, (x, y) => new
                        {
                            Id = x.Id,
                            Url = y.Url,
                            Parent = y.Parent,
                            MenuItem = y.MenuItem,
                            IsAuthorized = x.IsAuthorized,
                            PermissionRole = y.PermissionRole,
                            RoleId = x.RoleId,
                            ModuleId = y.ModuleId,
                            MenuId = y.Id,
                        }).ToList();
                }
                else
                {
                    output.model = context.Menus.Where(w => w.ModuleId == moduleId).Select(s => new
                    {
                        Id = 0,
                        Url = s.Url,
                        Parent = s.Parent,
                        MenuItem = s.MenuItem,
                        IsAuthorized = false,
                        PermissionRole = s.PermissionRole,
                        RoleId = role,
                        ModuleId = moduleId,
                        MenuId = s.Id,
                    }).ToList();
                }
                output.okay = true;
                output.message = string.Empty;
            }
            catch (Exception exception)
            {
                output.okay = false;
                output.model = null;
                output.message = exception.Message;
                logger.Error(exception.Message);
            }
            return output;
        }

        [HttpPost]
        [Route("setmenulist")]
        public object SetMenuList(List<MenuAuthentication> model)
        {
            Output output = new Output();
            try
            {
                foreach (var item in model)
                {
                    if (item.Id > 0)
                    {
                        var menu = context.MenuAuthentications.Find(item.Id);
                        menu.IsAuthorized = item.IsAuthorized;
                        context.SaveChanges();
                    }
                    else if (item.Id == 0)
                    {
                        context.MenuAuthentications.Add(item);
                        context.SaveChanges();
                    }

                }

                output.okay = true;
                output.message = string.Empty;
                output.model = model;

            }
            catch (Exception exception)
            {
                output.okay = false;
                output.model = null;
                output.message = exception.Message;
                logger.Error(exception.Message);
            }
            return output;
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (context != null)
                {
                    context.Dispose();
                }
            }
            base.Dispose(disposing);
        }

    }
}
